osconfig_tests/test_suites/package_management/package_management.go

```diff
@@ -112,44 +108,197 @@ func TestSuite(ctx context.Context, tswg *sync.WaitGroup, testSuites chan *junit
 }
 
+func runPackageInstallTest(ctx context.Context, testCase *junitxml.TestCase, testSetup *packageManagementTestSetup, logger *log.Logger) {
+
+	osConfig, err := osconfigserver.JsonToOsConfig(packageInstallTestOsConfigString, logger)
+
+	if err != nil {
+		testCase.WriteFailure("error while converting json to osconfig: \n%s\n", err)
+		return
+	}
+
+	oc := &osconfigserver.OsConfig{OsConfig: osConfig}
+
+	parent := fmt.Sprintf("projects/%s", testProject)
+	osconfigserver.CreateOsConfig(ctx, logger, oc, parent)
+
+	if err != nil {
+		testCase.WriteFailure("error while creating osconfig: \n%s\n", err)
+		return
+	}
+
+	defer cleanuposconfig(ctx, testCase, logger, oc)
+
+	assignment, err := osconfigserver.JsonToAssignment(packageInstallTestAssignmentString, logger)
+	if err != nil {
+		testCase.WriteFailure("error while converting json to assignment: \n%s\n", err)
+		return
+	}
+
+	assign := &osconfigserver.Assignment{Assignment: assignment}
+
+	res, err := osconfigserver.CreateAssignment(ctx, logger, assign, parent)
+	if err != nil {
+		testCase.WriteFailure("error while creating assignment: \n%s\n", err)
+	}
+	_ = res
+
+	defer cleanupassignment(ctx, testCase, logger, assign)
+
+	client, err := daisyCompute.NewClient(ctx)
+	if err != nil {
+		testCase.WriteFailure("error creating client: %v", err)
+		return
+	}
+
+	testCase.Logf("Creating instance with image %q", testSetup.image)
+	testSetup.name = fmt.Sprintf("osconfig-test-%s-%s", path.Base(testSetup.image), "packageinstalltest")
+
+	i := &api.Instance{
+		Name:        testSetup.name,
+		MachineType: fmt.Sprintf("projects/%s/zones/%s/machineTypes/n1-standard-1", testProject, testZone),
+		NetworkInterfaces: []*api.NetworkInterface{
+			&api.NetworkInterface{
+				Network: "global/networks/default",
+				AccessConfigs: []*api.AccessConfig{
+					&api.AccessConfig{
+						Type: "ONE_TO_ONE_NAT",
+					},
+				},
+			},
+		},
+		Metadata: &api.Metadata{
+			Items: []*api.MetadataItems{
+				testSetup.startup,
+			},
+		},
+		Disks: []*api.AttachedDisk{
+			&api.AttachedDisk{
+				AutoDelete: true,
+				Boot:       true,
+				InitializeParams: &api.AttachedDiskInitializeParams{
+					SourceImage: testSetup.image,
+				},
+			},
+		},
+		ServiceAccounts: []*api.ServiceAccount{
+			&api.ServiceAccount{
+				Email:  serviceAccountEmail,
+				Scopes: serviceAccountScopes,
+			},
+		},
+	}
+
+	inst, err := compute.CreateInstance(client, testProject, testZone, i)
+	if err != nil {
+		testCase.WriteFailure("Error creating instance: %v", err)
+		return
+	}
+	defer inst.Cleanup()
+
+	testCase.Logf("Waiting for agent install to complete")
+	if err := inst.WaitForSerialOutput("osconfig install done", 1, 5*time.Second, 5*time.Minute); err != nil {
+		testCase.WriteFailure("Error waiting for osconfig agent install: %v", err)
+		return
+	}
+
+	testCase.Logf("Agent installed successfully")
+
+	// allow agent to make the lookupconfig call and install the package
+	time.Sleep(1 * time.Minute)
```

Comment:
Sleeping for a fixed amount of time can lead to slow, flaky tests. Can this wait for serial out too?

Reply:
we can definitely make the test_runner wait on some kind of logs generated when the agent makes the lookupconfig call, for example, the agent logs the lookupconfig response.
However, I don't think waiting on a log that is generated by the agent would be robust. Lets imagine we decide to change the log line on which the test_runner depends, then we will have to change the test as well.
I would rather prefer the log line to be generated by the test_runner itself. Unfortunately, the test_runner does not have a way to know when the lookupconfig call is made by the agent.

I am thinking of rather than having a fixed and large sleeptime, I would change it to be a retry'ble command that would timeout after a certain number of tries. The waitforSerialOutput does the same thing.
Let me know if it is fine.

Reply:
had a discussion with @adjackura . we decided to use a startup script that keep writing to the serial console and then have the test case poll the serial output for a specific string.